api/v1alpha1/thanosreceive_types.go

```diff
@@ -0,0 +1,141 @@
+)
+
+// HashringAlgorithm represents the hashing algorithm to use.
+type HashringAlgorithm string
+
+const (
+	// AlgorithmKetama is the ketama hashing algorithm.
+	AlgorithmKetama HashringAlgorithm = "ketama"
+)
+
+// RouterSpec represents the configuration for the router
+type RouterSpec struct {
+	// Labels are additional labels to add to the router components.
+	// Labels set here will overwrite the labels inherited from the ThanosReceive object if they have the same key.
+	// +kubebuilder:validation:Optional
+	Labels map[string]string `json:"labels,omitempty"`
+	// Replicas is the number of router replicas.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Required
+	Replicas *int32 `json:"replicas,omitempty"`
+	// ReplicationFactor is the replication factor for the router.
+	// +kubebuilder:validation:Default=1
+	// +kubebuilder:validation:Enum=1;3;5
+	// +kubebuilder:validation:Required
+	ReplicationFactor *int32 `json:"replicationFactor,omitempty"`
+}
+
+// IngestorHashringSpec represents the configuration for a hashring to be used by the Thanos Receive StatefulSet.
+type IngestorHashringSpec struct {
+	// Name is the name of the hashring.
+	// Name will be used to generate the names for the resources created for the hashring.
+	// By default, Name will be used as a prefix with the ThanosReceive name as a suffix separated by a hyphen.
+	// In cases where that name does not match the pattern below, i.e. the name is not a valid DNS-1123 subdomain,
+	// the Name will be used as is and must be unique within the namespace.
+	// This field is immutable.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:Pattern=`^$|^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$`
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Value is immutable"
+	Name string `json:"name,omitempty"`
+	// Labels are additional labels to add to the hashring components.
+	// Labels set here will overwrite the labels inherited from the ThanosReceive object if they have the same key.
+	// +kubebuilder:validation:Optional
+	Labels map[string]string `json:"labels,omitempty"`
+	// Replicas is the number of replicas/members of the hashring to add to the Thanos Receive StatefulSet.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Required
+	Replicas *int32 `json:"replicas,omitempty"`
+	// Retention is the duration for which the Thanos Receive StatefulSet will retain data.
+	// +kubebuilder:default="2h"
+	// +kubebuilder:validation:Required
+	Retention Duration `json:"retention,omitempty"`
+	// ObjectStorageConfig is the secret's key that contains the object storage configuration.
+	// The secret needs to be in the same namespace as the ReceiveHashring object.
+	// +kubebuilder:validation:Required
+	ObjectStorageConfig corev1.SecretKeySelector `json:"objectStorageConfig"`
```

Comment:
@saswatamcode should we be able to specify this globally and just allow overrides down here?

Reply:
That might be a good idea actually. Will probably need to work it into ThanosStack cr